Dockerfile improvements

cameron1729 opened the pull request:
There are two parts to testing this PR:

- Testing the automatic generation of `maximalocal.mac.template` from the qtype_stack plugin
- Testing the ability to automatically build the container from the latest `qtype_stack` code, and the ability to specify a specific version to build from via CLI argument


### Testing automatic maximalocal template generation
1. Install `qtype_stack`  and its dependencies in a Moodle site (doesn't really matter which version of Moodle, just that you use the latest `qtype_stack`):
```
git clone  git@github.com:maths/moodle-qbehaviour_adaptivemultipart.git question/behaviour/adaptivemultipart
git clone  git@github.com:maths/moodle-qbehaviour_dfcbmexplicitvaildate.git question/behaviour/dfcbmexplicitvaildate
git clone git@github.com:maths/moodle-qbehaviour_dfexplicitvaildate.git question/behaviour/dfexplicitvaildate
git clone git@github.com:maths/moodle-qtype_stack.git question/type/stack
```
2. Clone this repo somewhere (doesn't matter where)
```
git clone git@github.com:catalyst/maximapool-docker.git
```
3. Check out the first commit from this repository (the one about fixing problems in the Dockerfile) and set up submodules:
```
cd maximapool-docker
git checkout dockerfile-improvements
git checkout dockerfile-improvements~2
git submodule update --init
```
4. Update the version of `qtype_stack` to match the version installed in Moodle:
```
cd assets/moodle-qtype_stack
git checkout master
```
5. Build the image (it will take a few minutes, and there will likely be some compilation warnings related to Maxima - they are safe to ignore):
```
cd -
docker build .
```
6. Once it finishes an image ID will be printed, e.g.,: `Successfully built [IMAGE ID]`
7. Start the container:
```
docker run --security-opt seccomp=unconfined -d \
                    --name MaximaPool \
                    -e "MAXIMAPOOL_ADMIN_PASSWORD=PASSWORD" \
                    -p "8765:8080" \
                    -v "$(pwd)/volumes/pool.conf.template:/opt/maximapool/pool.conf:ro" \
                    [IMAGE ID]
```
8. Check the output of `docker ps` until the container status shows as healthy
9. Obtain the IP address of the docker container:
```
sudo docker inspect -f "{{ .NetworkSettings.IPAddress }}" MaximaPool
```
10. On the Moodle site, browse to `admin/settings.php?section=qtypesettingstack`
11. Enter the following settings:
* Platform type: Server
* Maxima version: 5.41.0
* URL of the Maxima Pool: http://[IP ADDRESS FROM STEP 9]:8080/MaximaPool/MaximaPool
12. Save changes and then click the "healthcheck script" link (at the top of the page)
13. Note the "Maxima configuration file" section - prior to this PR, a manual step was required to update a Docker build asset with values from this section - in the commit we are currently testing, we expect to see unit test failures as we have not updated the build asset, we will then verify that the second commit in the PR automates this process and resolves the unit test failures
14. Scroll down to the bottom, verify graphs are visible and you see:
```
OK CAS returned data as expected.  You have a live connection to the CAS.
OK Correct and expected STACK-Maxima library version being used [VERSION NUMBER HERE].
```
15. Update `config.php` with the following lines:
```
define('QTYPE_STACK_TEST_CONFIG_PLATFORM', 'server');
define('QTYPE_STACK_TEST_CONFIG_MAXIMAVERSION', '5.41.0');
define('QTYPE_STACK_TEST_CONFIG_MAXIMACOMMAND', 'maxima --use-version=5.41.0');
define('QTYPE_STACK_TEST_CONFIG_MAXIMACOMMANDOPT', '');
define('QTYPE_STACK_TEST_CONFIG_MAXIMACOMMANDSERVER', 'http://[IP ADDRESS FROM STEP 9]:8080/MaximaPool/MaximaPool');
define('QTYPE_STACK_TEST_CONFIG_CASTIMEOUT', '5');
define('QTYPE_STACK_TEST_CONFIG_MAXIMALIBRARIES', 'stats, distrib, descriptive, simplex');
define('QTYPE_STACK_TEST_CONFIG_CASDEBUGGING', '0');
define('QTYPE_STACK_TEST_CONFIG_PLOTCOMMAND', '');
define('QTYPE_STACK_TEST_CONFIG_CASRESULTSCACHE', 'db');
define('QTYPE_STACK_TEST_CONFIG_CASPREPARSE', true);
```
15. Run the following test:
```
vendor/bin/phpunit --testsuite=qtype_stack_testsuite --filter=answertest_general_fixtures_test::test_answertest
```
16. **Verify** you get the following failure:
```
1) qtype_stack\answertest_general_fixtures_test::test_answertest with data set #1598 ('ATUnits( 120*kWh, 0.12*MWh) Options: 2', stdClass Object (...))
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'ATUnits_compatible_units (kg*m^2)/s^2.'
+'ATUnits_SB_no_units.'
```

This happens because when new unit definitions are added to the stack plugin (`Wh` in this case) Maxima needs its config updated (using the generated config mentioned in step 13). This is achieved in maximapool-docker by manually updating the  `assets/maximalocal.mac.template` file. If you go back to that config from step 13 and do a <kbd>Ctrl</kbd> + <kbd>f</kbd> for "Wh" you should find a match, but if you check the file ` assets/maximalocal.mac.template` in the maximapool-docker repo, you will see no instance of "Wh". This `maximalocal.mac.template` file is naively copied into the image without any update. The commit we are about to test automates the process of ensuring Maxima has an up-to-date config file when the image is built.
17. In the maximapool-docker repo, check out the commit which adds the ability to automatically generate the file:
```
git checkout dockerfile-improvements~1
```
18. Build the image again:
```
docker build .
```
19. Kill the old container and start the new one:
```
docker rm -f MaximaPool
docker run --security-opt seccomp=unconfined -d \
                    --name MaximaPool \
                    -e "MAXIMAPOOL_ADMIN_PASSWORD=PASSWORD" \
                    -p "8765:8080" \
                    -v "$(pwd)/volumes/pool.conf.template:/opt/maximapool/pool.conf:ro" \
                    [IMAGE ID]
```
20. Wait for the container status to come up healthy
21. Inspect the container's `maximalocal.mac.template` file:
```
docker exec MaximaPool bash -c "cat /opt/maximapool/maximalocal.mac.template" 
```
**Verify** that the `Wh` unit is present in the file.
22. Re-run the failing test (it might be necessary to update `config.php` to the IP of the new container, although in my case the IP didn't change):
```
vendor/bin/phpunit --testsuite=qtype_stack_testsuite --filter=answertest_general_fixtures_test::test_answertest
``` 
23. **Verify** the previous failure is not present. There may be another failure about `qtype_stack\answertest_general_fixtures_test::test_answertest with data set #1908` - this test seems to fail when run in isolation but passes when run as part of the entire testsuite - it's a separate issue that will need investigating and fixing in `qtype_stack`
24. Visit the healthcheck script again in Moodle and **verify** everything is still OK there

### Test building image automatically from GitHub
1. On a fresh Moodle site, set up an older version of `qtype_stack`:
```
git clone  git@github.com:maths/moodle-qbehaviour_adaptivemultipart.git question/behaviour/adaptivemultipart
git clone  git@github.com:maths/moodle-qbehaviour_dfcbmexplicitvaildate.git question/behaviour/dfcbmexplicitvaildate
git clone git@github.com:maths/moodle-qbehaviour_dfexplicitvaildate.git question/behaviour/dfexplicitvaildate
git clone git@github.com:maths/moodle-qtype_stack.git question/type/stack
cd question/type/stack
git checkout v4.4.1
``
2. In the maximapool-docker repo, check out the latest commit:
```
git checkout dockerfile-improvements
```
3. Build the image again:
```
docker build .
```
4. Kill the old container and start the new one:
```
docker rm -f MaximaPool
docker run --security-opt seccomp=unconfined -d \
                    --name MaximaPool \
                    -e "MAXIMAPOOL_ADMIN_PASSWORD=PASSWORD" \
                    -p "8765:8080" \
                    -v "$(pwd)/volumes/pool.conf.template:/opt/maximapool/pool.conf:ro" \
                    [IMAGE ID]
```
5. Wait for the container status to come up healthy
6. Repeat steps 9-12 from the previous testing section
7. Save changes and then click the "healthcheck script" link (at the top of the page)
8. Scroll to the bottom and check the server section, it should say:
```
OK CAS returned data as expected.  You have a live connection to the CAS.
FAILED The  version of the STACK-Maxima libraries being used (2023010400) does not  match what is expected (2022082900) by this version of the STACK  question type. Please rebuild the Maxima code on your MaximaPool server.
```

This is because the maxima pool image was built using the qtype_stack@master, and we have v4.4.1 locally in our moodle. We'll rebuild the image with a matching version next.
9. In the maximapool-docker repo, rebuild the image against qtype_stack@v4.4.1:
```
docker build --build-arg QTYPE_STACK_COMMIT="v4.4.1" .
```
10. Kill the old container and start the new one:
```
docker rm -f MaximaPool
docker run --security-opt seccomp=unconfined -d \
                    --name MaximaPool \
                    -e "MAXIMAPOOL_ADMIN_PASSWORD=PASSWORD" \
                    -p "8765:8080" \
                    -v "$(pwd)/volumes/pool.conf.template:/opt/maximapool/pool.conf:ro" \
                    [IMAGE ID]
```
11. Check the output of `docker ps` until the container status shows as healthy
12. Go back to the healthcheck script and press the "Clear the cache" button (down the bottom)
13. Check the server section of the healthcheck script, **verify** it says:
14. Scroll down to the bottom, verify graphs are visible and you see:
```
OK CAS returned data as expected.  You have a live connection to the CAS.
OK Correct and expected STACK-Maxima library version being used (2022082900).
```

cameron1729 committed:
Fix misc problems in Dockerfile

cameron1729 committed:
Add script to generate maximalocal.mac.template when using BUILD_FOR_MOODLE

The Maxima shipped with qtype_stack requires that the maximalocal.mac.template
updated whenver any change to units etc have been made. Previously updating
the template was a manual process that required visiting the qtype_stack
health check script, and copy pasting a specific seciton of the configuration
it generates in to the maximalocal.mac.template file.

This patch automates this process by grabbing the relevant script from the
qtype_stack plugin, hacking it to run, then using it to generate the new
configuration.

The static maximalocal.mac.template file is still left in this repo for the
case when NOT building for Moodle. In this case Maxima is taken from StackQuestion
and no modification to the maximalocal.mac.template is required.

cameron1729 commented on Dockerfile:
```diff
@@ -16,6 +16,18 @@ ENV MAXIMAPOOL=/opt/maximapool \
 
+RUN apt-get update \
+    && apt-get install -y \
+      openjdk-$(echo "$JAVA_VERSION" | cut -d '-' -f2 | cut -d '.' -f1)-jdk \
```
This is different from the old way (see below). I have no idea how the old way could have ever worked. The `JAVA_VERSION` environment variable looks like `jdk-11.0.18+10` and the `%%` thing extracts text from the start of the string up to the first instance of "not a string of digits in the range 0-9". For example:

```
$ BLAH="1234asdf"
$ echo  echo "${BLAH%%[!0-9]*}"
1234
```

What we want to extract from the `jdk-11.0.18+10`  string is the `11` part and there's just no way the old code can do that because it's specifically asking for things that are not numbers. Maybe there's a better way than what I've done here, but at least it works.

golenkovm commented:
As the template building process is now more automatic should we just remove `qtype_stack` asset and add a command to git clone the latest version?

cameron1729 commented:
> As the template building process is now more automatic should we just remove `qtype_stack` asset and add a command to git clone the latest version?

Is there any downside to just leaving it as a git module? The instructions already say to run `git clone --recurse-submodules`  when you want to build an image (although we should probably update it so the instructions reference this repo instead of the archived one).

Another option could be removing the asset, and modifying the Dockerfile to clone the `qtype_stack` repo. By default it would just get the latest master, but we could add an arg to allow specifying a specific commit.

cameron1729 committed:
Remove question type assets

Instead of building from a pre-bundled qtype_stack, the Dockerfile
now pulls the latest qtype_stack from GitHub, and gets files from
there.

Two Dockerfile build arguments are now provided:

- QTYPE_STACK_COMMIT: To specify an exact commit to checkout (defaults to "master")
- QTYPE_STACK_REMOTE: To specify where to clone qtype_stack from (defaults to the canonical repo)

The ability to build against `assStackQuestion` has been removed.

cameron1729 commented:
@golenkovm based on our discussion yesterday I've added a commit which changes the build process to grab the latest `qtype_stack` code from GitHub instead of using a pre-bundled asset.

Two additional build arguments have been added:

- QTYPE_STACK_COMMIT: To specify an exact commit (or branch, tag, etc) to checkout (defaults to "master")
- QTYPE_STACK_REMOTE: To specify where to clone qtype_stack from (defaults to the [canonical repo](https://github.com/maths/moodle-qtype_stack))

Please have a look and let me know what you think :)

golenkovm commented:
Thanks for implementing this @cameron1729 
The PR looks good to me. Happy to have it merged.
Cheers
Misha

cameron1729 commented:
Closing this one as we now use goemaxima.
